Automatic refunds

0nko opened the pull request:
Fixes #1478. This PR brings the automatic refund by utilising the payment gateway endpoint. When the app is started (or a site changes), all payment gateways for the site are loaded. During a refund the selected gateway is checked, whether it supports automatic refunds.

The feature flag for issuing refunds has been removed.

<img src="https://user-images.githubusercontent.com/1522856/67111831-22c2d500-f1d6-11e9-9d5c-a2d80e302de5.png" width="300" />

<img src="https://user-images.githubusercontent.com/1522856/67111842-28b8b600-f1d6-11e9-97b9-b01fe19df05e.png" width="300" />

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

JavonDavis commented:
Hey! Bumping this to `3.0`, if you'd like to get it in for this release please merge this into release/2.9 and then DM me to release a new Beta version.

0nko commented:
Thanks @anitaa1990! I've addressed all the comments and the PR's ready for another round.

> The payment gateway endpoint is not called when the user logs in to the app for the first time.

Huh, didn't notice it while testing.. Thanks! I've changed it as you suggested.

> Soft keypad is not closed when clicking on the bottom tab from the Issue Refund by amount screen.

I've hidden the bottom navigation so this won't be an issue anymore.

> The refund list in Order Detail is not automatically updated when a refund is issued.

The refund flow does send the result back to the order detail to refresh the screen but it seems it takes some time for the data to appear on the server. I added a 1s delay which seems to do the trick.

> I'm not able to reproduce this issue but I got this crash when clicking on the Issue Refund button in Order detail.

This used to happen if you navigate from the refund flow by tapping on the bottom navigation, so it shouldn't happen anymore.

> Just mentioning it here so it doesn't get missed but one small suggestion would be to validate and register the events added to the PR before approval.

Will do, thanks!

anitaa1990 commented:
@0nko, thanks for the changes. The changes look good 👍  

I'm just facing some minor and weird issues when testing.

- A couple of tests in the `OrderListPresenterTest` class are failing.

<img width="1356"  src="https://user-images.githubusercontent.com/22608780/67356943-7c7e2300-f579-11e9-914e-2b109442b9ad.png">
 
- The previously refunded summary disappears when clicking on Settings and coming back to the screen. Note that this only happens when `Do not keep activities` is enabled.

<img width="300" src="https://user-images.githubusercontent.com/22608780/67357091-d0890780-f579-11e9-8de7-4a082082da84.gif">

- When clicking on the `Refund` button and clicking back before the refund is processed:
  - The refund list is not updated in Order detail, even though I got an email saying the refund was completed.
  - When trying to refund again for a different order, after entering the amount, a snackbar is displayed stating `Your refund was successfully submitted`, when the refund was not processed. I _think_ the snackbar is displayed from the previous order refund here.

<img width="300" src="https://user-images.githubusercontent.com/22608780/67357396-bc91d580-f57a-11e9-88cd-3b1c18e2d6d0.gif">

There are also some merge conflicts in the PR.

0nko commented:
> A couple of tests in the OrderListPresenterTest class are failing.

This should be fixed now.

> The previously refunded summary disappears when clicking on Settings and coming back to the screen. Note that this only happens when Do not keep activities is enabled.

The reason for this is because state is not being preserved by the ViewModels. I started implementing the ViewModel SavedState library but it turns out to be a pretty massive change. I'll put it in a separate PR.

> The refund list is not updated in Order detail, even though I got an email saying the refund was completed.

This is a bummer. Because the refund creation result is being returned as LiveData to a fragment, which is now dismissed, I couldn't think of a way to fix this easily... 😕I'll keep it in mind and try to come up with something in a subsequent PR.

> When trying to refund again for a different order, after entering the amount, a snackbar is displayed stating Your refund was successfully submitted, when the refund was not processed. I think the snackbar is displayed from the previous order refund here.

I've reset the SingleLiveEvents so at least this is not happening anymore.

Thanks for the thorough testing
